### PR TITLE
Fix 1569 winning hero with no troops shall retreat

### PR DESCRIPTION
**Pull request by vmarkovtsev**
Try to fix http://bugs.vcmi.eu/view.php?id=1569
The following assumptions were made:
1) Winning hero with no army still gains XP, new levels, etc.
2) Raised dead with Necromancery skill are added before the check, so hero remains alive in such cases

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
Is everything's OK now?

File: server/CGameHandler.cpp [hunk 1]
```diff
@@ -734,7 +734,25 @@ void CGameHandler::battleAfterLevelUp( const BattleResult &result )
...
+		if(const CGHeroInstance *another =  getPlayer(finishingBattle->loser)->availableHeroes.at(0))
+			sah.hid[1] = another->subID;
+		else
+			sah.hid[1] = -1;
+
+		sendAndApply(&sah);
+	}
+	if(finishingBattle->winnerHero && finishingBattle->winnerHero->stacks.empty())
+	{
+		RemoveObject ro(finishingBattle->winnerHero->id);
```
Comment (ArseniyShestakov) [inline server/CGameHandler.cpp]: Now it's will crash there on RemoveObject if both heroes died in battle. It's easier to reproduce with spell like Armageddon while two not too powerful heroes are fighting.

Comment (vmarkovtsev) [inline server/CGameHandler.cpp]: Got it. I know how to fix

**Comment by vmarkovtsev** [comment]
Actually, when Armageddon kills everybody, there is a separate memory corruption somewhere in runBattle(). In that case it crashes far before battleAfterLevelUp(), in
```
battleQuery->result = *battleResult.data;
```
And by trial and effort I found that battleResult.data->artifacts segfaults during copy. Besides, I get
```
There is already set result?
```
log message. Sometimes, it crashes in malloc() in runBattle(), so something goes wrong and unrelated to this PR. After manually disabling battleResult.data->artifacts copying with some dark magic code, the battle successfully ended and both heroes died.

The only thing I had to add is the special handling of draws.

**Comment by vmarkovtsev** [comment]
I think this can be merged and a separate issue for the described crash on MacOSX should be composed.

**Comment by ArseniyShestakov** [comment]
I don't think crash you get is something OS X exclusive.
I think there was problem with crashes on draw in past.

**Comment by vmarkovtsev** [comment]
Gotcha! I am a debug king!

```
static void giveExp(BattleResult &r)
{
	r.exp[0] = 0;
	r.exp[1] = 0;
	for(auto i = r.casualties[!r.winner].begin(); i!=r.casualties[!r.winner].end(); i++)
	{
		r.exp[r.winner] += VLC->creh->creatures.at(i->first)->valOfBonuses(Bonus::STACK_HEALTH) * i->second;
	}
}
```
When we have a draw, r.winner equals to 2, while exp is int[2] => memory corruption inside std::set ("artifacts" field go right after "exp").
Besides, I found a race inside CGameHandler::setBattleResult() leading to double write.

**Comment by alexvins** [comment]
Nice catch! This should resolve http://bugs.vcmi.eu/view.php?id=1566

**Comment by ArseniyShestakov** [comment]
Now it's works fine and I found few problems related to draw where it not handled properly so I'll likely fix them or report on bug tracker.

**Comment by ArseniyShestakov** [comment]
Also it's funny, but I tested H3 and I can't reproduce that behaviour at all. If I lose hero while having summoned creatures It's never appear in tavern: not on siege (neutral, or hero with town), not in battle with neutral creatures.

**Comment by vmarkovtsev** [comment]
I think winning hero should appear in tavern after all - having won, he could virtually retreat while summoned creatures still exist.

**Comment by ArseniyShestakov** [comment]
I agree and there no reason to not include that by default. We have several options that alter original game mechanics: dwelling population reset, some changes in pathfinding and may be something else too.

Though if that behaviour actually not present in H3 then we may want to make it optional.

**Comment by vmarkovtsev** [comment]
I see. Will add the corresponding switch after getting familiar how these options are worked with.

**Comment by ArseniyShestakov** [comment]
You can just add it to "hardcodedFeatures".

**Comment by vmarkovtsev** [comment]
I did what you said, but eventually it breaks the save format, so I made it a separate PR so that this one can be merged at once and the other can be deferred.

**Pull request merged**